module/extension/Director/src/director/challenge_priority.cpp

```diff
@@ -58,23 +59,23 @@ namespace module::extension {
         }
 
         // Function to get the priorities of the ancestors of this task
-        auto get_ancestor_priorities = [this](const std::shared_ptr<const BehaviourTask>& task) {
+        auto get_ancestor_priorities = [this](const std::shared_ptr<BehaviourTask>& task) {
             // We are our first ancestor
             std::vector<TaskPriority> ancestors;
 
             // The task might be a root task, in which case we won't have any provider
             // In that case we set the type to nullptr_t to indicate that we are a root task
             ancestors.emplace_back(
-                providers.count(task->requester_id) != 0 ? providers[task->requester_id]->type : typeid(nullptr_t),
+                providers.count(task->requester_id) != 0 ? providers.at(task->requester_id)->type : typeid(nullptr_t),
                 task->priority,
                 task->optional);
 
-            // Loop up through the providers until we reach a point where a task was emitted by a non provider
-            for (auto t = task; providers.count(t->requester_id) != 0;) {
+            // Loop up through the providers until we reach a point where a task was emitted by a root provider
+            for (auto t = task; providers.at(t->requester_id)->classification != Provider::Classification::ROOT;) {
 
                 // Get the provider that emitted this task, and from that the provider group
-                auto provider = providers[t->requester_id];
-                auto& group   = groups[provider->type];
+                auto provider = providers.at(t->requester_id);
```

Comment:
Or is there a reason we want to copy this?
```suggestion
                auto& provider = providers.at(t->requester_id);
```

Reply:
It's a shared pointer, I generally take those by copy even when I could take them by reference